v2/bundle/jwtbundle/bundle.go

```diff
@@ -1,73 +1,164 @@
+		}
+	}
+
+	return bundle, nil
 }
 
 // TrustDomain returns the trust domain the bundle belongs to.
 func (b *Bundle) TrustDomain() spiffeid.TrustDomain {
-	panic("not implemented")
+	b.mtx.RLock()
+	defer b.mtx.RUnlock()
+
+	return b.trustDomain
 }
 
 // JWTKeys returns the JWT keys in the bundle, keyed by key ID.
 func (b *Bundle) JWTKeys() map[string]crypto.PublicKey {
-	panic("not implemented")
+	b.mtx.RLock()
+	defer b.mtx.RUnlock()
+
+	return b.jwtSigningKeys
 }
 
 // FindJWTKey finds the JWT key with the given key id from the bundle. If the key
 // is found, it is returned and the boolean is true. Otherwise, the returned
 // value is nil and the boolean is false.
 func (b *Bundle) FindJWTKey(keyID string) (crypto.PublicKey, bool) {
-	panic("not implemented")
+	b.mtx.RLock()
+	defer b.mtx.RUnlock()
+
+	if jwtKey, ok := b.jwtSigningKeys[keyID]; ok {
+		return jwtKey, true
+	}
+	return nil, false
+}
+
+// HasJWTKey returns true if the bundle has a JWT key with the given key id.
+func (b *Bundle) HasJWTKey(keyID string) bool {
+	b.mtx.RLock()
+	defer b.mtx.RUnlock()
+
+	_, ok := b.jwtSigningKeys[keyID]
+	return ok
 }
 
 // AddJWTKey adds a JWT key to the bundle. If a JWT key already exists
-// under the given key ID, it is replaced.
-func (b *Bundle) AddJWTKey(keyID string, key crypto.PublicKey) {
-	panic("not implemented")
+// under the given key ID, it is replaced. A key ID must be specified.
+func (b *Bundle) AddJWTKey(keyID string, key crypto.PublicKey) error {
+	if keyID == "" {
+		return errors.New("missing key ID")
+	}
+
+	b.mtx.Lock()
+	defer b.mtx.Unlock()
+
+	b.jwtSigningKeys[keyID] = key
+	return nil
 }
 
 // RemoveJWTKey removes the JWT key identified by the key ID from the bundle.
 func (b *Bundle) RemoveJWTKey(keyID string) {
-	panic("not implemented")
+	b.mtx.Lock()
+	defer b.mtx.Unlock()
+
+	delete(b.jwtSigningKeys, keyID)
 }
 
 // Marshal marshals the JWT bundle into a standard RFC 7517 JWKS document. The
 // JWKS does not contain any SPIFFE-specific parameters.
 func (b *Bundle) Marshal() ([]byte, error) {
-	panic("not implemented")
+	b.mtx.RLock()
+	defer b.mtx.RUnlock()
+
+	jwks := jose.JSONWebKeySet{}
+	for keyID, jwtSigningKey := range b.jwtSigningKeys {
+		if keyID == "" {
+			return nil, errors.New("missing key ID")
+		}
+		jwks.Keys = append(jwks.Keys, jose.JSONWebKey{
+			Key:   jwtSigningKey,
+			KeyID: keyID,
+		})
+	}
+
+	return json.MarshalIndent(jwks, "", "    ")
 }
 
 // GetJWTBundleForTrustDomain returns the JWT bundle of the given trust domain.
 // It implements the Source interface. It will fail if called with a trust
 // domain other than the one the bundle belongs to.
 func (b *Bundle) GetJWTBundleForTrustDomain(trustDomain spiffeid.TrustDomain) (*Bundle, error) {
-	panic("not implemented")
+	b.mtx.RLock()
+	defer b.mtx.RUnlock()
+
+	if b.trustDomain != trustDomain {
+		return nil, fmt.Errorf("this bundle does not belong to trust domain %q", trustDomain)
```

Comment:
I wonder if we should make this error more generic.... the `Set` type could use the same verbage.
```suggestion
		return nil, fmt.Errorf("no JWT bundle for trust domain %q", trustDomain)
```